### ISSUE TITLE
Connection to borker takes about one second, should be quicker?

### ISSUE THREAD

**Issue by lifeisafractal** [eclipse/paho.mqtt.cpp#365]
I'm new to the Paho C++ client, but I'm having an issue when using the `async_client`. When I connect and then wait on the contention, it takes right around 1 second to connect. This feels like it's taking to long for this simple operation.

```c++
mqtt::async_clinet client("localhost", "clientid")
auto start = std::chrono::high_resolution_clock::now().time_since_epoch();
client->connect()->wait();
auto end = std::chrono::high_resolution_clock::now().time_since_epoch();
printf("connect took: %d milliseconds\n", (end - start) / 1000000);
```
I'm getting right around 1000ms give or take a few on every run. Also, it is suspiciously right around 1s every time which makes me wonder if there is some fixed delay or timeout happening in the connection process that doesn't need to be there (just speculation).

In contrast, when I use the Paho python client it takes between 10-15ms to connect.

```python
client = mqtt.Client(protocol=mqtt.MQTTv311)
start = time.time()
client.connect("localhost")
end = time.time()
print(f"Duration {end - start}")
```
I'm getting right around 12ms with a few ms variance on this code which makes sense since this would result in just a few context switches between the broker and the python client.

For a broker I'm using Mosquitto v2.0.10 running on `localhost`.

I assume I'm missing something fundamental, but can't seem to figure out why this might be the case after reviewing the documentation. Any help is appreciated!

**Comment by fpagliughi** [comment]
The C++ library is a fairly thin wrapper around the Paho C library. I personally have not seen this behavior on the connect, but saw the same thing with _disconnect()_. A pause that was almost exactly 1sec:
https://github.com/eclipse/paho.mqtt.c/issues/889

Is the problem consistent? Or intermittent?
What system are you running the client on? Linux? Windows?

You can turn on runtime tracing for the C lib with environment variables:
```
export MQTT_C_CLIENT_TRACE=ON
export MQTT_C_CLIENT_TRACE_LEVEL=MEDIUM
```
as described here: https://github.com/eclipse/paho.mqtt.c#runtime-tracing

See if you can capture a log showing the delay and we can try to find it.

**Comment by lifeisafractal** [comment]
> Is the problem consistent? Or intermittent?

So far this has been very consistent for me. I don't believe I've ever seen it connect quickly with the C++ lib on my development system.

> What system are you running the client on? Linux? Windows?

I'm on Linux (XUbuntu 21.04).

My (possibly incorrect) understanding is that the Paho Python libs are also a wrapper on top of the Paho C libs which is why I pointed out the difference in my above example. Let me know if I'm wrong there, this would be useful to know if I'm barking up the wrong tree there.

I'll try turning on runtime tracing and report back with the results. I wont have time for a little bit to check that out unfortunately (maybe the next day or two).

Thanks for the quick and helpful reply Frank!

**Comment by fpagliughi** [comment]
No. The Python Paho library is _not_ a wrapper for C. It's completely written in Python.

**Comment by fpagliughi** [comment]
But I knew the connect problem sounded familiar! Someone reported the same thing in the Rust library, and that is also a wrapper around the C lib. 
https://github.com/eclipse/paho.mqtt.rust/issues/133

I'll try to recreate the problem and then push the issue upstream.

**Comment by lifeisafractal** [comment]
So I was able to run and get a trace and I got pretty interesting results. I'm getting similar behavior to the issue in the Rust wrapper issue you pointed to above. The first client seems to be able to connect quickly, and the second hangs up for a second.

Here is a snippet of the trace around the second client connecting. It looks like a call int `Thread_wait_cond` is timing out at 1 second and causing this delay. I'm having a bit of trouble making sense of the trace and I think the output from the two different clients is getting interleaved making it hard to tell whats going on. I'll keep poking at this, bust wanted to post what I found in case that shakes something loose.

```
20210916 134753.903 (3628025664) (0)< MQTTAsync_connect:908 (0)
20210916 134753.903 (3628009024)  (1)< Thread_wait_cond:429 (0)
20210916 134753.903 (3628009024)  (1)> MQTTAsync_checkTimeouts:1659
20210916 134753.903 (3628009024)  (1)< MQTTAsync_checkTimeouts:1747
20210916 134753.903 (3628009024)  (1)> MQTTAsync_processCommand:1180
20210916 134753.903 Connecting to serverURI localhost with MQTT version 4
20210916 134753.903 (3628009024)   (2)> MQTTProtocol_connect:218
20210916 134753.903 (3628009024)    (3)> MQTTProtocol_addressPort:60
20210916 134753.903 (3628009024)    (3)< MQTTProtocol_addressPort:92
20210916 134753.903 (3628009024)    (3)> Socket_new:696
20210916 134753.903 New socket 4 for localhost, port 1883
20210916 134753.903 (3628009024)     (4)> Socket_addSocket:174
20210916 134753.903 (3628009024)      (5)> Socket_setnonblocking:86
20210916 134753.903 (3628009024)      (5)< Socket_setnonblocking:91 (0)
20210916 134753.903 (3628009024)     (4)< Socket_addSocket:209 (0)
20210916 134753.903 Connect pending
20210916 134753.903 (3628009024)    (3)< Socket_new:852 (115)
20210916 134753.903 (3628009024)   (2)< MQTTProtocol_connect:340 (115)
20210916 134753.903 (3628009024)   (2)> MQTTProperties_free:401
20210916 134753.903 (3628009024)   (2)< MQTTProperties_free:424
20210916 134753.903 (3628009024)  (1)< MQTTAsync_processCommand:1571 (1)
20210916 134753.903 (3628009024)  (1)> Thread_wait_cond:414
20210916 134753.903 (3628009024)  (1)< Thread_wait_cond:429 (110)
20210916 134753.903 (3628009024)  (1)> MQTTAsync_checkTimeouts:1659
20210916 134753.903 (3628009024)  (1)< MQTTAsync_checkTimeouts:1747
20210916 134753.903 (3628009024)  (1)> Thread_wait_cond:414
20210916 134753.903 (3619616320)    (3)> Socket_continueWrites:995
20210916 134753.903 (3619616320)    (3)< Socket_continueWrites:1025 (0)
20210916 134753.903 (3619616320)    (3)> isReady:226
20210916 134753.903 (3619616320)    (3)< isReady:231 (0)
20210916 134753.903 (3619616320)    (3)> isReady:226
20210916 134753.903 (3619616320)    (3)< isReady:231 (1)
20210916 134754.904 (3619616320)   (2)< Socket_getReadySocket:328 (4)
20210916 134754.904 m->c->connect_state = 1
```

**Comment by lifeisafractal** [comment]
Just a thought, since this seems to be an issue in the underlying C lib rather than either the Rust or C++ wrappers, should I open up an issue in that project to consolidate discussion there?

**Comment by lifeisafractal** [comment]
Another bit of info, it looks like the condition variable is global rather than unique to a particular client instance. My guess is that there is some interaction/contention going on when having multiple connections in the same process. It seems to me that different clients shouldn't need to be synchronized in anyway between each other, and each client instance should hold it's own mutexes and condition variables. I'm not familiar with the internals of the code however, so I could be wrong.

https://github.com/eclipse/paho.mqtt.c/blob/master/src/MQTTAsync.c#L268

**Comment by fpagliughi** [comment]
This appears to be resolved with the Paho C lib v1.3.10.
With both the C++ and Rust wrapper libraries.

**Comment by fpagliughi** [comment]
This seems to have been resolved with the update to the C library. Anyone can feel free to reopen this or create a new Issue if anything similar is happening with C library newer than v1.3.10.

**Comment by JinyongJeong-TIM** [comment]
@fpagliughi In the latest code, I have similar issue. If I try to connect multiple async client in the situation that there is one connected client, connection take about 2.1 second.  This is the log

```
[2024-09-27 05:29:32.588] [info] [mqtt_connection_test.cpp:40] Connection time: 2.103402082
[2024-09-27 05:29:34.692] [info] [mqtt_connection_test.cpp:40] Connection time: 2.103529591
[2024-09-27 05:29:36.795] [info] [mqtt_connection_test.cpp:40] Connection time: 2.103572912
[2024-09-27 05:29:38.898] [info] [mqtt_connection_test.cpp:40] Connection time: 2.102895892
[2024-09-27 05:29:41.001] [info] [mqtt_connection_test.cpp:40] Connection time: 2.102235044
[2024-09-27 05:29:41.001] [info] [mqtt_connection_test.cpp:42] Total connection time: 10.515635521
```
I think this issue is not resolved yet.
